docs/source/export.rst

```diff
@@ -1,10 +1,445 @@
+* ``exported_program.equality_constraints`` describes which dimensions are
+  required to be equal. Since we specified in the constraints that the first
+  dimension of each argument is equivalent,
+  (``dynamic_dim(example_args[0], 0) == dynamic_dim(example_args[1], 0)``),
+  we see in the equality constraints the tuple specifying that ``arg4_1``
+  dimension 0 and ``arg5_1`` dimension 0 are equal.
+
+
+Serialization
+^^^^^^^^^^^^^
+
+To save the ``ExportedProgram``, users can use the :func:`torch.export.save` and
+:func:`torch.export.load` APIs. It will serialize the ``ExportedProgram`` to a
+Python dataclass, and then save it as a JSON format. The ``state_dict`` will be
+saved to a file separate from the serialized ``ExportedProgram``.
+
+A convention is to save the ``ExportedProgram`` using a ``.pt2`` file extension.
+
+An example:
+
+::
+
+    import torch
+    import io
+
+    class MyModule(torch.nn.Module):
+        def forward(self, x):
+            return x + 10
+
+    exported_program = torch.export.export(MyModule(), torch.randn(5))
+
+    torch.export.save(exported_program, 'exported_program.pt2')
+    saved_exported_program = torch.export.save('exported_program.pt2')
+
+
+Specialization
+^^^^^^^^^^^^^^
+
+Input shapes
+~~~~~~~~~~~~
+
+As mentioned before, by default, ``torch.export`` will trace the program
+specializing on the input tensors' shapes, unless a dimension is specified as
+dynamic via the :func:`torch.export.dynamic_dim` API. This means that if there
+exists shape-dependent control flow, ``torch.export`` will specialize on the
+branch that is being taken with the given sample inputs. For example:
+
+::
+
+    import torch
+    from torch.export import export
+
+    def fn(x):
+        if x.shape[0] > 5:
+            return x + 1
+        else:
+            return x - 1
+
+    example_inputs = (torch.rand(10, 2),)
+    exported_program = export(fn, example_inputs)
+    print(exported_program)
+    """
+    ExportedProgram:
+        class GraphModule(torch.nn.Module):
+            def forward(self, arg0_1: f32[10, 2]):
+                add: f32[10, 2] = torch.ops.aten.add.Tensor(arg0_1, 1);
+                return (add,)
+    """
+
+The conditional of (``x.shape[0] > 5``) does not appear in the
+``ExportedProgram`` because the example inputs have the static
+shape of (10, 2). Since ``torch.export`` specializes on the inputs' static
+shapes, the else branch (``x - 1``) will never be reached. To preserve the dynamic
+branching behavior based on the shape of a tensor in the traced graph,
+:func:`torch.export.dynamic_dim` will need to be used to specify the dimension
+of the input tensor (``x.shape[0]``) to be dynamic, and the source code will
+need to be :ref:`rewritten <Data/Shape-Dependent Control Flow>`.
+
+Non-tensor inputs
+~~~~~~~~~~~~~~~~~
+
+``torch.export`` also specializes the traced graph based on the values of inputs
+that are not ``torch.Tensor``, such as ``int``, ``float``, ``bool``, and ``str``.
+For example:
```

Comment:
Maybe add a note saying that we will likely change this in the near future to not specialize on values of primitive type inputs.